components/brave_wallet/browser/zcash/rust/orchard_shard_tree_impl.cc

```diff
@@ -0,0 +1,347 @@
+  return ::brave_wallet::OrchardShardAddress{addr.level, addr.index};
+}
+
+ShardTreeAddress ToRust(const ::brave_wallet::OrchardShardAddress& addr) {
+  return ShardTreeAddress{addr.level, addr.index};
+}
+
+ShardTreeCap ToRust(::brave_wallet::OrchardShardTreeCap& shard_store_cap) {
+  ::rust::Vec<uint8_t> data;
+  data.reserve(shard_store_cap.size());
+  base::ranges::copy(shard_store_cap, std::back_inserter(data));
+  return ShardTreeCap{std::move(data)};
+}
+
+::brave_wallet::OrchardShardTreeCap FromRust(const ShardTreeCap& cap) {
+  ::brave_wallet::OrchardShardTreeCap shard_store_cap;
+  shard_store_cap.reserve(cap.data.size());
+  base::ranges::copy(cap.data, std::back_inserter(shard_store_cap));
+  return shard_store_cap;
+}
+
+::brave_wallet::OrchardShard FromRust(const ShardTreeShard& tree) {
+  std::optional<OrchardShardRootHash> shard_root_hash;
+  if (!tree.hash.empty()) {
+    CHECK_EQ(kOrchardShardTreeHashSize, tree.hash.size());
+    OrchardShardRootHash hash_value;
+    base::ranges::copy(tree.hash, hash_value.begin());
+    shard_root_hash = hash_value;
+  }
+
+  std::vector<uint8_t> data;
+  data.reserve(tree.data.size());
+  base::ranges::copy(tree.data, std::back_inserter(data));
+
+  return ::brave_wallet::OrchardShard(FromRust(tree.address), shard_root_hash,
+                                      std::move(data));
+}
+
+ShardTreeShard ToRust(const ::brave_wallet::OrchardShard& tree) {
+  ::rust::Vec<uint8_t> data;
+  data.reserve(tree.shard_data.size());
+  base::ranges::copy(tree.shard_data, std::back_inserter(data));
+
+  ::rust::Vec<uint8_t> hash;
+  if (tree.root_hash) {
+    base::ranges::copy(tree.root_hash.value(), std::back_inserter(hash));
+  }
+  return ShardTreeShard{ToRust(tree.address), std::move(hash), std::move(data)};
+}
+
+ShardTreeCheckpoint ToRust(
+    const ::brave_wallet::OrchardCheckpoint& checkpoint) {
+  ::rust::Vec<uint32_t> marks_removed;
+  base::ranges::copy(checkpoint.marks_removed,
+                     std::back_inserter(marks_removed));
+  return ShardTreeCheckpoint{!checkpoint.tree_state_position.has_value(),
+                             checkpoint.tree_state_position.value_or(0),
+                             marks_removed};
+}
+
+ShardTreeCheckpointBundle ToRust(
+    const ::brave_wallet::OrchardCheckpointBundle& checkpoint_bundle) {
+  return ShardTreeCheckpointBundle(checkpoint_bundle.checkpoint_id,
+                                   ToRust(checkpoint_bundle.checkpoint));
+}
+
+::brave_wallet::OrchardCheckpoint FromRust(
+    const ShardTreeCheckpoint& checkpoint) {
+  CheckpointTreeState checkpoint_tree_state = std::nullopt;
+  if (!checkpoint.empty) {
+    checkpoint_tree_state = checkpoint.position;
+  }
+  return ::brave_wallet::OrchardCheckpoint{
+      checkpoint_tree_state,
+      std::vector<uint32_t>(checkpoint.mark_removed.begin(),
+                            checkpoint.mark_removed.end())};
+}
+
+CxxOrchardShardTreeDelegate::CxxOrchardShardTreeDelegate(
```

Comment:
again, please put each class in it's own file unless an internal class or something that doesn't have a public header

Reply:
Done